mon/OSDMonitor: add warning when deploying crimson OSDs without enabling flags

gitkenan opened the pull request:
This is just a partial attempt for review at the moment. 

----------------------------------


Attempt at first PR which adds to the code for a good-first-issue enhancement mentioned here: https://tracker.ceph.com/issues/66610 ... I'm expecting to have missed a lot here or made errors so any cc is appreciated!

In essence, the idea is to print a warning when OSDs are deployed on a Crimson build without the necessary flags set; in particular, the flag related to osds.





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

anthonyeleven commented:
Jenkins retest this plesase

docs lgtm

gitkenan commented:
> Jenkins retest this plesase
> 
> docs lgtm

ty! :) I repeated the jenkins tests, unfortunately still getting similar results with a standard C++ lib failure, I guess I'll wait for now and retry in a day or so, thanks again!

Matan-B reviewed (changes requested):
The existing code warns when disallowing boot of crimson-osd. However, as the tracker attached describes, the single log (info) warning may be not enough to understand why no OSDs are "up".
I think that we should be more explicit and perhaps raise a health message that would be seen easily when monitoring the cluster's status.

---

Creating a new pool without `osd_pool_default_crimson` is also possible by specifying `crimson` on creation. See:
```
    bool crimson = cmd_getval_or<bool>(cmdmap, "crimson", false) ||
      cct->_conf.get_val<bool>("osd_pool_default_crimson");
```
As not using `osd_pool_default_crimson` probably worth an additional warning, I think that the more important case we should handle first is better explaining why no OSDs are coming up.

gitkenan commented on src/mon/OSDMonitor.cc:
```diff
@@ -3520,6 +3521,10 @@ bool OSDMonitor::preprocess_boot(MonOpRequestRef op)
       }
+      // warn if crimson flags are not correctly set for OSD deployment
+      if (!osdmap.is_osd_pool_default_crimson_set(cct)) {
+      ss << "WARNING: Booting crimson-osd without osd_pool_default_crimson set.";
```
> I'm not sure that `ss` will be used later on. See `reply_no_propose:` or `reply:` for `ss` usages.

Thanks for this - will work accordingly. I noticed recently that in OSDMap.cc for example, there's a lot of use of 'ss' and no 'reply' or 'reply_no_purpose' to convert that into rs in this file. Would you like me to raise a separate PR to add that feature into the files where there's a lot of ss usage, and replace ss with rs in files where there's a little rs usage?

Matan-B replied:
> As a side question: I noticed recently that in src/osd/OSDMap.cc for example, there's a lot of use of 'ss' and no sign of any 'reply' or 'reply_no_purpose' to convert that into rs in this file. 

I meant that `reply_no_propose/reply` are 2 examples in which `OSDMonitor.cc` uses `ss`. It's not a general convention, `src/osd/OSDMap.cc` probably has other usages as well.

gitkenan replied:
Oh okay, thanks!

gitkenan commented on src/cephadm/tests/test_deploy.py:
```diff
@@ -558,7 +558,7 @@ def test_deploy_smb_container(cephadm_fs, funkypatch):
         runfile_lines = f.read().splitlines()
     assert 'podman' in runfile_lines[-1]
-    assert runfile_lines[-1].endswith('quay.io/essembee/samba-server:latest --samba-debug-level=6 run smbd')
+    assert runfile_lines[-1].endswith('quay.io/essembee/samba-server:latest run smbd')
```
sorry this wasn't me, I'll remove this soon if I get the chance

gitkenan commented on src/cephadm/cephadmlib/daemons/smb.py:
```diff
@@ -279,7 +279,6 @@ def validate(self) -> None:
             custom_dns=custom_dns,
             domain_member=Features.DOMAIN.value in instance_features,
             clustered=Features.CLUSTERED.value in instance_features,
-            samba_debug_level=6,
```
this wasn't me, sneaky commit

gitkenan commented on src/mon/OSDMonitor.cc:
```diff
@@ -3520,6 +3520,16 @@ bool OSDMonitor::preprocess_boot(MonOpRequestRef op)
+          << "data loss.  Its usage should be restricted to "
+	  << "testing and development.";
+        goto ignore;
+      }
```
I decided to go with the info error for this, to follow the error before it, given that we're now passing a health warning.

gitkenan reviewed (commented):
I've changed the approach a little here, since there's two separate mechanisms for handling the flag for OSDs not being set, I've removed the ostringstream ss approach for detecting only the flag being disabled, and for the health warning it should be very clear to the monitor.

I've still not tested this myself yet manually, I'll do that once I'm able hopefully, which shouldn't be too long. I pushed this as it's my current approach.

gitkenan commented on src/osd/OSDMap.cc:
```diff
@@ -57,6 +57,7 @@ using ceph::decode;
 using ceph::encode;
 using ceph::Formatter;
 
+#define dout_context g_ceph_context
```
eek sorry, forgot to remove this from some work I cancelled. please ignore for now.

gitkenan reviewed (commented):
Extra info about my code changes, in particular some mistakes that I hope to fix soon when on my computer again.

gitkenan commented on src/osd/OSDMap.cc:
```diff
@@ -7330,6 +7330,15 @@ void OSDMap::check_health(CephContext *cct,
       auto& d = checks->add("OSD_DOWN", HEALTH_WARN, ss.str(),
 			    down_in_osds.size());
+      bool crimson_build_but_flags_disabled = cmd_getval_or<bool>(cmdmap, "crimson", false) &&
+               !cct->_conf->get_val<bool>("osd_pool_default_crimson");
```
My thinking here was that I want to raise the warning if we're running a crimson build, but the flag's not set. Maybe this "cmdmap" crimson value is not the one I want? I saw other options like `experimental_features_enabled`, and/or 'ifdef SEASTAR', but they seemed far fetched...

gitkenan reviewed (commented):
re-thinking this after re-reading your comment...

gitkenan commented on src/osd/OSDMap.cc:
```diff
@@ -7330,6 +7333,20 @@ void OSDMap::check_health(CephContext *cct,
+      bool crimson_pool_is_configured = cmd_getval_or<bool>(cmdmap, "crimson", false) ||
+               cct->_conf->get_val<bool>("osd_pool_default_crimson");
+      #endif
+      if (crimson_build && !crimson_pool_is_configured) {
```
I tried to do this with one boolean, which is still possible, but I made it as two for the moment for two main reasons:

- clarity of code and to communicate my reasoning
- when using only one boolean, I found it hard to keep a descriptive name and also have it make sense in its default value if crimson is not enabled (I would have had to set the variable as true when WITH_SEASTAR isn't true, which, with the naming, made no sense

gitkenan reviewed (commented):
Hopefully this will be the last review comments on my side for now - thanks!

athanatos commented on src/osd/OSDMap.cc:
```diff
@@ -57,6 +57,9 @@ using ceph::decode;
 
+#ifndef WITH_SEASTAR
+  #include "common/Cond.h"
+#endif
```
OSDMap.h has a WITH_SEASTAR ifdef for the smart pointer type.  Other than that, I really don't want OSDMap to have compile-time differences for WITH_SEASTAR.

athanatos commented on src/osd/OSDMap.cc:
```diff
@@ -7330,6 +7333,20 @@ void OSDMap::check_health(CephContext *cct,
+      bool crimson_pool_is_configured = cmd_getval_or<bool>(cmdmap, "crimson", false) ||
+               cct->_conf->get_val<bool>("osd_pool_default_crimson");
+      #endif
+      if (crimson_build && !crimson_pool_is_configured) {
```
I'm not really fond of adding crimson specific build behavior to OSDMap beyond the smart pointer ifdef in the header.  It's *very* important that most of the methods in OSDMap behave identically across the mons, osds, and clients so we don't really want build time behavior changes.

crimson_build doesn't seem to be declared if !WITH_SEASTAR, so it probably won't compile?

Finally, this method will actually be invoked on the monitor, not the crimson-osd, so WITH_SEASTAR will be false regardless.

athanatos replied:
Also, I think it's the allow_crimson flag on the OSDMap that we care about, not the osd_pool_default_crimson config flag.

I don't think you can actually emit the warning from the OSDMap.  If OSDMap::allow_crimson (OSDMap::get_allow_crimson) is false, we simply ignore the boot message from the OSD and its metadata doesn't get added to the map.  See OSDMonitor::preprocess_boot.

gitkenan replied:
Thanks for all of this, it's going to help me a lot with understanding ceph in general-I used this PR as a way to learn more about C++, ceph and crimson and the whole development cycle, and comments like this are a huge help so thanks again

athanatos replied:
Yep, no worries.  Ideally, we'd just trigger the health warning from preprocess_boot, but I don't actually remember how the health warning mechanism works.  Feel free to ping me if you can't find a good way to do it and I'll take a look.

gitkenan commented:
I've tried to take all of the above comments to account, with two HEALTH_WARNs; one in the case of pools not being able to be created properly as in what was mentioned in @Matan-B 's comment, and another one specifically for when not all flags are set. That way, when the experimental data corrupting flag isn't set for example, only the latter warning is given. I guess this may not be what is required, but I guess it's better to have more caution and warnings than less in this scenario. Let me know if you want modifications and I'll try to get to it ASAP. Thanks!

gitkenan commented:
Also I should mention that I'm _almost_ into the sepia lab, but until then, I haven't actually tried to compile or test this yet- so this review request is more about my logic.

gitkenan commented on src/mon/OSDMonitor.cc:
```diff
@@ -3510,15 +3511,33 @@ bool OSDMonitor::preprocess_boot(MonOpRequestRef op)
-	  << "data loss.  Its usage should be restricted to "
-	  << "testing and development.";
-	goto ignore;
+      cmdmap_t cmdmap;
```
A little bit doubtful about how I'm initialising the cmdmap here and if it's sufficient to get the crimson attribute.

gitkenan commented:
So I've ran some manual testing and despite the fact that it compiles successfully now, I'm not seeing the results that I want. I suspect this is related to how I am initialising the 'cmdmap' in the file changes; I've put a little note there too to point at exactly what I mean. Here are my testing details (I consulted with @Matan-B about this before I tried this approach of editing the vstart script):

after making the changes in the PR, I did the following:

```
./do_cmake.sh -DWITH_SEASTAR=ON
cd build
ninja -l 40 -j 20
ninja install
```

Then I edited the ../src/vstart.sh script to comment out the three flags in question, here is the diff:
```
[root@cephbase ceph]# git diff src/vstart.sh

diff --git a/src/vstart.sh b/src/vstart.sh
index a462a9ae60a..5d95323e71a 100755
--- a/src/vstart.sh
+++ b/src/vstart.sh
@@ -827,7 +827,7 @@ prepare_conf() {
         plugin dir = $CEPH_LIB
         run dir = $CEPH_OUT_DIR
         crash dir = $CEPH_OUT_DIR
-        enable experimental unrecoverable data corrupting features = *
+        # enable experimental unrecoverable data corrupting features = *
         osd_crush_chooseleaf_type = 0
         debug asok assert abort = true
         $(format_conf "${msgr_conf}")
@@ -988,7 +988,7 @@ EOF

     if [ "$crimson" -eq 1 ]; then
         wconf <<EOF
-        osd pool default crimson = true
+        # osd pool default crimson = true
 EOF
     fi

@@ -1125,7 +1125,8 @@ EOF
     done

     if [ "$crimson" -eq 1 ]; then
-        $CEPH_BIN/ceph osd set-allow-crimson --yes-i-really-mean-it
+        :
+       # $CEPH_BIN/ceph osd set-allow-crimson --yes-i-really-mean-it
     fi

     if [ -n "$require_osd_and_client_version" ]; then
```
then I saved the script and ran the following:

```
MGR=1 MON=1 OSD=1 MDS=0 RGW=0 ../src/vstart.sh   --without-dashboard --bluestore --crimson   --redirect-output
source ./vstart_environment.sh
```
then, I see:

```
[root@cephbase build]# ceph status
  cluster:
    id:     098525f0-4c7d-46b7-ab49-55d27bc00a76
    health: HEALTH_WARN
            Reduced data availability: 33 pgs stale
            2 pool(s) have no replicas configured

  services:
    mon: 1 daemons, quorum a (age 6m)
    mgr: x(active, since 6m)
    osd: 1 osds: 1 up (since 17s), 1 in (since 62m)

  data:
    pools:   2 pools, 33 pgs
    objects: 2 objects, 449 KiB
    usage:   1.0 GiB used, 100 GiB / 101 GiB avail
    pgs:     33 stale+active+clean

[root@cephbase build]# ceph health detail
HEALTH_WARN 2 pool(s) have no replicas configured
[WRN] POOL_NO_REDUNDANCY: 2 pool(s) have no replicas configured
    pool 'rbd' has no replicas configured
    pool '.mgr' has no replicas configured

```
But since I have (in my mind) started a crimson build but without setting these essential 3 flags, I should have triggered my health errors and I want to see something like this:
```

[root@cephbase build]# ceph status
  cluster:
    id:     098525f0-4c7d-46b7-ab49-55d27bc00a76
    health: HEALTH_WARN
            Mandatory crimson flags are not all set.
            Crimson is not configured for making pools.

  services:
    mon: 1 daemons, quorum a (age 6m)
    mgr: x(active, since 6m)
    osd: 1 osds: 1 up (since 17s), 1 in (since 62m)

  data:
    pools:   2 pools, 33 pgs
    objects: 2 objects, 449 KiB
    usage:   1.0 GiB used, 100 GiB / 101 GiB avail
    pgs:     33 stale+active+clean

[root@cephbase build]# ceph health detail
[WRN] CRIMSON_FLAGS_OFF: At least one of the three mandatory crimson flags are not set.
If OSDs are 'in' but 'down', this will most likely be the reason why.
See the 'enabling crimson' section in the crimson documentation
for more information.

[WRN] CRIMSON_OSDS_DOWN: Crimson OSDs are down because relevant flags are not enabled.
If OSDs are 'in' but 'down', this will most likely be the reason why.
See the 'enabling crimson' section in the crimson documentation
for more information.
```

athanatos commented:
@gitkenan Apologies, I didn't quite make it to this today -- I should be able to find time tomorrow.

gitkenan commented on src/mon/OSDMonitor.cc:
```diff
@@ -3510,15 +3511,33 @@ bool OSDMonitor::preprocess_boot(MonOpRequestRef op)
+                                  osdmap.get_allow_crimson() &&
+                                  g_ceph_context->check_experimental_feature_enabled("crimson");
+      if (!all_mandatory_crimson_flags_are_set){
+        ostringstream ss, ss2;
```
I've left this without a goto reply_no_propose at the end of the block because there's no such thing in this file

Matan-B commented on src/mon/OSDMonitor.cc:
```diff
@@ -3510,15 +3511,33 @@ bool OSDMonitor::preprocess_boot(MonOpRequestRef op)
     const auto &otype = type_iter->second;
     // m->metadata["osd_type"] must be "crimson", classic doesn't send osd_type
     if (otype == "crimson") {
-      if (!osdmap.get_allow_crimson()) {
-	mon.clog->info()
-	  << "Disallowing boot of crimson-osd without allow_crimson "
-	  << "OSDMap flag.  Run ceph osd set-allow-crimson to set "
-	  << "allow_crimson flag.  Note that crimson-osd is "
-	  << "considered unstable and may result in crashes or "
-	  << "data loss.  Its usage should be restricted to "
-	  << "testing and development.";
-	goto ignore;
```
Let's keep it

Matan-B reviewed (commented):
Expanding my previous comment https://github.com/ceph/ceph/pull/58879#pullrequestreview-2207120123.

Crimson has 3 flags which are needed to be set:

1)  'enable_experimental_unrecoverable_data_corrupting_features crimson

    In case unrecoverable_data_corrupting_features is not enabled with crimson, we block the command used:
    ```
    bool experimental_enabled =
      g_ceph_context->check_experimental_feature_enabled("crimson");
    if (!sure || !experimental_enabled) {
      ss << "This command will allow usage of crimson-osd osd daemons.  "
	 << "crimson-osd is not considered stable and will likely cause "
	 << "crashes or data corruption.  At this time, crimson-osd is mainly "
	 << "useful for performance evaluation, testing, and development.  "
	 << "If you are sure, add --yes-i-really-mean-it and add 'crimson' to "
	 << "the experimental features config.  This setting is irrevocable.";
      err = -EPERM;
      goto reply_no_propose;
    }
    ```

2) osd_pool_default_crimson

    In case default Crimson pool creation is not enabled. We are still able to create Crimson pools by specifically mentioning:
    ```

    bool crimson = cmd_getval_or<bool>(cmdmap, "crimson", false) ||
      cct->_conf.get_val<bool>("osd_pool_default_crimson");
    ```

    Crimson OSD won't instantiate PGs from non-Crimson pools:
    ```
            } else if (!pool->is_crimson()) {
            DEBUG("ignoring pgid {}, pool lacks crimson flag", pgid);
    ```

3) set-allow-crimson

    Without this flag, OSDs won't go up.
    
---

My suggestion is as follows:

As 1 is already quite verbose and informative we can skip adding any additional warning to not setting it.

For 2, I think that we should add the health warning **once** pg creation is actually blocked.
This should also help with avoiding the cmdmap usage.

For 3, same as 2, let's add the health warning only once OSD boot is blocked. (as implemented here).

`OSDMonitor::preprocess_boot` shouldn't be responsible for checking all the flags, only the mon one `osd_pool_default_crimson` is relevant to it. 

Lastly, for the question regarding not seeing any health warning. I would suggest to begin with a simple (testing) health warning.
Make sure that you are able to set/clear it properly and then continue to the actual intended behavior. That way you'll be able to spot issues and bugs earlier.

Feel free to ping once ready/not clear! :)

gitkenan commented on src/osd/OSDMap.h:
```diff
@@ -880,6 +880,12 @@ class OSDMap {
+    cct->_conf->get_value("osd_pool_default_crimson", value);
+    return value;
+  }
+
```
this should be ignored for now

gitkenan commented on src/mon/OSDMonitor.cc:
```diff
@@ -3519,6 +3520,7 @@ bool OSDMonitor::preprocess_boot(MonOpRequestRef op)
 	  << "data loss.  Its usage should be restricted to "
 	  << "testing and development.";
 	goto ignore;
+    // Other health warning has to go here
```
ignore for now (not implemented yet)

anthonyeleven commented on src/common/options/mon.yaml.in:
```diff
@@ -495,6 +495,13 @@ options:
+- name: mon_warn_on_pool_pg_creation_blocked
+  type: bool
+  level: dev
+  desc: issue CRIMSON_FLAG_OFF warning if necessery osd_pool_default_crimson flag is not set, as PG creation won't happen without it
```
s/necessery/the/

gitkenan commented:
> Expanding my previous comment [#58879 (review)](https://github.com/ceph/ceph/pull/58879#pullrequestreview-2207120123).
> 
> Crimson has 3 flags which are needed to be set:
> 
> 1. 'enable_experimental_unrecoverable_data_corrupting_features crimson
>    In case unrecoverable_data_corrupting_features is not enabled with crimson, we block the command used:
>    ```
>    bool experimental_enabled =
>      g_ceph_context->check_experimental_feature_enabled("crimson");
>    if (!sure || !experimental_enabled) {
>      ss << "This command will allow usage of crimson-osd osd daemons.  "
>     << "crimson-osd is not considered stable and will likely cause "
>     << "crashes or data corruption.  At this time, crimson-osd is mainly "
>     << "useful for performance evaluation, testing, and development.  "
>     << "If you are sure, add --yes-i-really-mean-it and add 'crimson' to "
>     << "the experimental features config.  This setting is irrevocable.";
>      err = -EPERM;
>      goto reply_no_propose;
>    }
>    ```
> 2. osd_pool_default_crimson
>    In case default Crimson pool creation is not enabled. We are still able to create Crimson pools by specifically mentioning:
>    ```
>    
>    bool crimson = cmd_getval_or<bool>(cmdmap, "crimson", false) ||
>      cct->_conf.get_val<bool>("osd_pool_default_crimson");
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    Crimson OSD won't instantiate PGs from non-Crimson pools:
>    ```
>            } else if (!pool->is_crimson()) {
>            DEBUG("ignoring pgid {}, pool lacks crimson flag", pgid);
>    ```
> 3. set-allow-crimson
>    Without this flag, OSDs won't go up.
> 
> My suggestion is as follows:
> 
> As 1 is already quite verbose and informative we can skip adding any additional warning to not setting it.
> 
> For 2, I think that we should add the health warning **once** pg creation is actually blocked. This should also help with avoiding the cmdmap usage.
> 
> For 3, same as 2, let's add the health warning only once OSD boot is blocked. (as implemented here).
> 
> `OSDMonitor::preprocess_boot` shouldn't be responsible for checking all the flags, only the mon one `osd_pool_default_crimson` is relevant to it.
> 
> Lastly, for the question regarding not seeing any health warning. I would suggest to begin with a simple and unconditional (testing) health warning. Make sure that you are able to set/clear it properly and then continue to the actual intended behavior. That way you'll be able to spot issues and bugs earlier.
> 
> Feel free to ping once ready/not clear! :)

Hi there! 

So I think I've been struggling a bit to understand/meet the requirements - I don't think I understand the architecture enough to know where these health warnings should go with the least amount of hassle. If you could give me a set place to put them, that would help a lot, so that I won't waste your time bouncing back another review. For now, I've pushed something for us to look at - just the health warning for set-allow-crimson! :)

For the set-allow-crimson flag, I thought of putting it in HealthMonitor.cc, around here:

```cpp

bool HealthMonitor::check_leader_health()
{
  dout(20) << __func__ << dendl;
  bool changed = false;

  // prune quorum_health
  {
    auto& qset = mon.get_quorum();
    auto p = quorum_checks.begin();
    while (p != quorum_checks.end()) {
      if (qset.count(p->first) == 0) {
	p = quorum_checks.erase(p);
	changed = true;
      } else {
	++p;
      }
    }
  }

  health_check_map_t next;

 // DAEMON_OLD_VERSION
  if (g_conf().get_val<bool>("mon_warn_on_older_version")) {
    check_for_older_version(&next);
  }
  // MON_DOWN
  check_for_mon_down(&next);
  // MON_CLOCK_SKEW
  check_for_clock_skew(&next);
  // MON_MSGR2_NOT_ENABLED
  if (g_conf().get_val<bool>("mon_warn_on_msgr2_not_enabled")) {
    check_if_msgr2_enabled(&next);
  }

 +++ check_if_set_allow_crimson_flag_is_set($next) <<< New function here

  if (next != leader_checks) {
    changed = true;
    leader_checks = next;
  }
  return changed;
}
```

However, importing the necessary metadata for crimson checks in this file seemed unnecessarily complex. Even more so was the case for the shards_... file where PGs are being blocked. 

I thought of putting it here, in OSDMonitor::preprocess_boot:

```cpp
if (!osdmap.get_allow_crimson()) {
	mon.clog->info()
	  << "Disallowing boot of crimson-osd without allow_crimson "
	  << "OSDMap flag.  Run ceph osd set-allow-crimson to set "
	  << "allow_crimson flag.  Note that crimson-osd is "
	  << "considered unstable and may result in crashes or "
	  << "data loss.  Its usage should be restricted to "
	  << "testing and development.";
	goto ignore;
      }
```

But it was mentioned above that this part isn't relevant for raising a health warning for this flag (and that the osd_pool_default_crimson flag is the only one which is mon-related). 

I managed to get a set-allow-crimson warning working for a function encode_... in the OsdMon, but I had doubts about its relevancy because I'd have to re-init the crimson boolean where it wasn't being used, so it felt out of place. I have done it anyway and pushed the change, which had the following result:

```bash
[root@cephbase build]# ceph status
2024-11-30T15:10:18.844+0000 7f4b7cd85640 -1 WARNING: all dangerous and experimental features are enabled.
2024-11-30T15:10:18.860+0000 7f4b7cd85640 -1 WARNING: all dangerous and experimental features are enabled.
  cluster:
    id:     55750f97-1e97-4d1d-9b82-18908f7d9572
    health: HEALTH_WARN
            1 Crimson OSDs are IN but not UP: 'set-allow-crimson' flag not set
            Reduced data availability: 32 pgs inactive
            1 pool(s) have no replicas configured

  services:
    mon: 1 daemons, quorum a (age 61m)
    mgr: x(active, since 60m)
    osd: 1 osds: 0 up, 1 in (since 60m)

  data:
    pools:   1 pools, 32 pgs
    objects: 0 objects, 0 B
    usage:   0 B used, 0 B / 0 B avail
    pgs:     100.000% pgs unknown
             32 unknown
```
However, I wasn't sure how good of an idea it was to be extracting osd_type metadata in a place where it's not already being used. 

It does however work:

```bash
[root@cephbase build]# ceph osd set-allow-crimson --yes-i-really-mean-it
2024-12-01T10:48:28.891+0000 7f267a8ce640 -1 WARNING: all dangerous and experimental features are enabled.
2024-12-01T10:48:28.917+0000 7f267a8ce640 -1 WARNING: all dangerous and experimental features are enabled.
2024-12-01T10:48:29.195+0000 7f31ec7de640 -1 WARNING: experimental feature 'crimson' is enabled
Please be aware that this feature is experimental, untested,
unsupported, and may result in data corruption, data loss,
and/or irreparable damage to your cluster.  Do not use
feature with important data.

[root@cephbase build]# ceph status
2024-12-01T10:48:31.601+0000 7f4d4a9e3640 -1 WARNING: all dangerous and experimental features are enabled.
2024-12-01T10:48:31.616+0000 7f4d4a9e3640 -1 WARNING: all dangerous and experimental features are enabled.
  cluster:
    id:     12ffe038-a9a8-46a4-8ca0-2a7e48d2d06b
    health: HEALTH_WARN
            Reduced data availability: 32 pgs inactive
            1 pool(s) have no replicas configured
 
  services:
    mon: 1 daemons, quorum a (age 2m)
    mgr: x(active, since 2m)
    osd: 1 osds: 0 up, 1 in (since 2m)
 
  data:
    pools:   1 pools, 32 pgs
    objects: 0 objects, 0 B
    usage:   0 B used, 0 B / 0 B avail
    pgs:     100.000% pgs unknown
             32 unknown
 
[root@cephbase build]# 
```

Since I felt a bit lost as to where to put these, if you have an idea off of the top of your head as to where I can implement both flags, that'd be ideal (I wasn't sure by 'preprocess_boot' is relevant to osd_pool_default_crimson you meant that the health warning should go in there or not - a bit more info will hopefully help me from wasting more of your time!).

Matan-B commented on src/mon/OSDMonitor.cc:
```diff
@@ -2061,8 +2061,37 @@ void OSDMonitor::encode_pending(MonitorDBStore::TransactionRef t)
   // health
   health_check_map_t next;
+
+  if (auto type_iter = m->metadata.find("osd_type");
+      type_iter != m->metadata.end()) {
+    const auto &otype = type_iter->second;
+    // m->metadata["osd_type"] must be "crimson", classic doesn't send osd_type
+    if (otype == "crimson" && !tmp.get_allow_crimson()) {        
```
We handle `MOSDBoot` messages in `OSDMonitor::preprocess_boot`, we should probably try to reuse this logic and set the warning from there. As mentioned, we shouldn't handle these messages in `encode_pending`.

Getting a health warning from `preprocess_boot` would be tricky with the current warning design we have. It seems that we encode health checks only from `encode_pending`. Optimally, we would have been able to add the health check in `OSDMap::check_health`. However, OSDMap current structure is not able to know about our warning.
We could  try adding the warning information to OSDMap (similiarly to what we did with `OSDMap::allow_crimson`), though I don't think that it's worth the overhead of storing that in every OSDMap shared.

If a health warning would end up being too tricky to implement, we could perhaps reply to the OSD (MOSDBootReply?) so that the OSD will go down once with a proper abort message once this message is received?
@athanatos, do you have a better suggestion?

Matan-B reviewed (commented):
```
    health: HEALTH_WARN
            1 Crimson OSDs are IN but not UP: 'set-allow-crimson' flag not set
```

Great work on finding a way to set and clear the warning! Left a comment regarding the next steps :)
